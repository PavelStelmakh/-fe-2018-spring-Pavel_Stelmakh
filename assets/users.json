[
  {
    "id": 1,
    "name": "Pasha",
    "age": "19",
    "password": "1111",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  },
  {
    "id": 2,
    "name": "Sasha",
    "age": "24",
    "password": "2222",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  },
  {
    "id": 3,
    "name": "Masha",
    "age": "20",
    "password": "3333",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  }
]