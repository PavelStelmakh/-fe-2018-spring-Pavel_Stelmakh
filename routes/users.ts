import { Router, Response, Request, NextFunction } from 'express';
const router = Router();
import { User } from '../models/User';
const loginRoute = require('./login');
const users = require('../models/Users');
import { getUserByToken } from '../shared/getUserByToken';

router.use('/login', loginRoute);

router.use('*', (req: Request, res: Response, next: NextFunction) => {
    const data = getUserByToken(req);

    if (data.user) {
        next();
    } else {
        res.status(401);
        res.end();
    }
});

router.get('/users', (_req: Request, res: Response) => {
    res.status(200).send(users.users());
});

router.get('/users/search/:name', (req: Request, res: Response) => {
    try {
        const name: number = req.params['name'];
        const searchUsers: User[] = users.search(name);
        setTimeout(() => {
            if (searchUsers.length > 0) {
                res.status(200).send(searchUsers);
            } else {
                res.status(400);
                res.end();
            }
        }, 1000);
    }
    catch (Error) {
        res.status(418);
        res.end();
    }
});

router.get('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const user: User | undefined = users.user(id);
        if (user) {
            setTimeout(() => res.status(200).send(user), 3000);
        } else {
            res.status(400);
            res.end();
        }
    }
    catch (Error) {
        res.status(418);
        res.end();
    }
});

router.post('/users/add', (req: Request, res: Response) => {
    try {
        const user: User = users.add(req.body);
        res.status(200).send(user);
    }
    catch (Error) {
        res.status(418);
    }
    res.end();
});

router.put('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const success: boolean = users.update(id, req.body);
        if (success) {
            res.status(200);
        } else {
            res.status(418);
        }
    }
    catch (Error) {
        res.status(500);
    }
    res.end();
});

router.delete('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const success: boolean = users.delete(id);
        if (success) {
            res.status(200);
        } else {
            res.status(418);
        }
    }
    catch (Error) {
        res.status(500);
    }
    res.end();
});

module.exports = router;