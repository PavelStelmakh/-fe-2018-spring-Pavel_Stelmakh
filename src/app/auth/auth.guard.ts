import { Injectable } from '@angular/core';
import { CanLoad, ActivatedRouteSnapshot, RouterStateSnapshot, Router, Route, CanActivate } from '@angular/router';
import { Observable, of } from 'rxjs';
import { mergeMap, take, catchError } from 'rxjs/operators';
import { Store } from '@ngrx/store';
import * as profileAction from '../actions/profile.action';
import * as profileReducer from '../reducers/profile.reduser';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class AuthGuard implements CanLoad, CanActivate {
  constructor(
    private router: Router,
    private authService: AuthService,
    private store: Store<profileReducer.State>
    ) {}

    canLoad(route: Route): Observable<boolean> | Promise<boolean> | boolean {
      const {path} = route;
      return this.authService.checkLogin().pipe(
        take(1),
        mergeMap(result => {
          if (result.body['role'] === 'admin') {
            this.store.dispatch(new profileAction.SetIdAction(+result.body['id']));
            if (path === 'admin') {
              return of(true);
            } else {
              this.router.navigate(['/admin']);
              return of(false);
            }
          } else {
            if (path === 'user') {
              return of(true);
            } else {
              this.router.navigate(['/user']);
              return of(false);
            }
          }
        }),
        catchError(error => {
          this.router.navigate(['/login']);
          return of(false);
        })
      );
  }

  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
    return this.authService.checkLogin().pipe(
    take(1),
    mergeMap(result => {
      this.store.dispatch(new profileAction.SetIdAction(+result.body['id']));
      switch (state.url) {
        case '/login':
        case '/login/recovery':
          this.router.navigate([`/${result.body['role']}`]);
          return of(false);
        default: return of(true);
      }   
    }),
    catchError(error => {
      switch (state.url) {
        case '/login':
        case '/login/recovery': 
          return of(true);
        default: 
          this.router.navigate(['/login']);
          return of(false);
        }
      })
    );
  }

}
