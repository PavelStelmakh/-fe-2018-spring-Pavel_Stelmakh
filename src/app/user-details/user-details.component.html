<div appSpinner>
    <div></div>
    <div></div>
    <div></div>
    <div></div>
    <div></div>
    <div></div>
    <div></div>
</div>
<div [class.tabs]="true" (click)="onToggle($event)">
    <div [ngClass]="{selectTabFirst: true, select: selectFirst}">{{'user-details.tabOut' | translate}}</div>
    <div [ngClass]="{selectTabSecond: true, select: !selectFirst}">{{'user-details.tabIn' | translate}}</div>
</div>
<div *ngIf="selectFirst; then tabOut else tabIn"></div>
<ng-template #tabOut><app-output-form></app-output-form></ng-template>
<ng-template #tabIn><app-user-form (update)="reverseSelect()"></app-user-form></ng-template>
<input type="button" value="{{'user-details.logout' | translate}}" (click)="logout()">