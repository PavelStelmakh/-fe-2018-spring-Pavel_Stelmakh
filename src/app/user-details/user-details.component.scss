
@keyframes detailsAnimation {
    from {
        top: 0;
        background-color: rgba(0, 0, 0, 0);
        box-shadow: 0 0 0px 0px rgba(0, 0, 0, 0);
    }
    to {
        top: 20vh;
        background-color: rgba(0, 0, 0, 0.7);
        box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.5);
    }
}

:host {    
    text-align: center;
    display: block;
    width: 500px;
    background-color: rgba(0, 0, 0, 0.7);
    position: absolute;
    top: 20vh;
    left: calc(50vw - 250px);
    box-shadow: 0 0 10px 3px rgba(0,0,0,0.5);
    animation: detailsAnimation 1s;

    .tabsView {
        padding: 10px 0;
    }
}

input {
    margin: 30px 0;
}

.tabs div {
    color: #ccc;
    position: relative;
    background-color: rgba(102, 102, 102, 0.5);
    display: inline-block;
    width: 250px;
    height: 40px;
    line-height: 40px;
    cursor: pointer;
    transition: left 200ms ease-in 200ms, right 200ms ease-in;
    
    &:hover {
        background-color: rgba(102, 102, 102, 0.7);
    }
}

.selectTabFirst {
    left: 0;
}

.selectTabSecond {
    right: 0;
}

.selectTabFirst.select {
    left: -5px;
}

.selectTabSecond.select {
    right: -5px;
}
