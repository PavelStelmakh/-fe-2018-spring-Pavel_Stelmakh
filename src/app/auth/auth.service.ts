import { Injectable, Inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { BASE_URL } from '../url.service';
import { SignIn } from 'models/SignIn';
import { Recovery } from 'models/Recovery';
import { tap } from 'rxjs/operators';
import { UsersService } from '../users.service';

@Injectable({
  providedIn: 'root'
})
export class AuthService {

  constructor(
    private http: HttpClient, 
    @Inject(BASE_URL) private url: string,
    private user: UsersService
    ) { }

  signIn(userData: SignIn) {
    return this.http.post<SignIn>(`${this.url}/login`, userData, {
      headers: new HttpHeaders({
        'Content-Type':  'application/json'
      }),
      observe: 'response',
      withCredentials: true
    });
  }

  recovery(userData: Recovery) {
    return this.http.post<SignIn>(`${this.url}/login/recovery`, userData, {
      headers: new HttpHeaders({
        'Content-Type':  'application/json'
      }),
      observe: 'response'
    });
  }

  checkLogin() {
    return this.http.get(`${this.url}/login`, {
      observe: 'response',
      withCredentials: true
    })
    .pipe(
      tap(result => {
        if (result.status === 200) {
          this.user.setId(+result.body['id']);
        }
      })
    );
  }

  logout() {
    return this.http.get(`${this.url}/login/logout`, {
      observe: 'response',
      withCredentials: true
    });
  }

}
