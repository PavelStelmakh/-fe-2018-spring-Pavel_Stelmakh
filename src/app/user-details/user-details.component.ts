import { Component, OnInit, Input } from '@angular/core';
import { AuthService } from '../auth/auth.service';
import { Router } from '@angular/router';
import { UsersService } from '../users.service';
import { Store } from '@ngrx/store';
import * as profileReducer from '../reducers/profile.reduser';
import * as profileAction from '../actions/profile.action';

@Component({
  selector: 'app-user-details',
  templateUrl: './user-details.component.html',
  styleUrls: ['./user-details.component.scss']
})
export class UserDetailsComponent implements OnInit {
  select: number;
  isEditUsers: boolean;
  @Input() role: string;

  constructor(
    private auth: AuthService, 
    private router: Router,
    private userService: UsersService,
    private store: Store<profileReducer.State>
    ) {}

  ngOnInit() {
    this.isEditUsers = false;
    this.select = 0;
    this.store.dispatch(new profileAction.LoadProfileAction());
  }

  logout() {
    this.auth.logout().subscribe(result => {
      if (result.status === 200) {
        this.userService.logout();
        this.store.dispatch(new profileAction.SetIdAction(null));
        this.router.navigate(['/login']);
      }
    },
    error => {
      alert('Неизвестная ошибка');
    });
  }

  onToggle(tab: number){
    this.isEditUsers = false;
    this.select = tab;
  }

}
