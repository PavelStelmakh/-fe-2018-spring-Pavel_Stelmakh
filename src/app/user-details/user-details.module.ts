import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { TranslateModule, TranslateLoader, TranslateService } from '@ngx-translate/core';
import { Store, select } from '@ngrx/store';

import { HttpLoaderFactory } from '../shared/HttpLoaderFactory';
import { UserListModule } from '../user-list/user-list.module';
import { OutputFormComponent } from './output-form/output-form.component';
import { TreeModule } from '../tree/tree.module';
import { UserDetailsComponent } from './user-details.component';
import { BASE_URL } from '../url.service';
import { UserFormComponent } from './user-form/user-form.component';
import { InvalidInputComponent } from './invalid-input/invalid-input.component';
import { HttpClientModule, HttpClient } from '@angular/common/http';
import * as language from '../reducers/language.reducer';

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    HttpClientModule,
    TranslateModule.forChild({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient, BASE_URL]
      },
      isolate: true
    }),
    UserListModule,
    TreeModule
  ],
  declarations: [
    UserFormComponent,
    InvalidInputComponent,
    UserDetailsComponent,
    OutputFormComponent,
  ],
  providers: [
    {provide: BASE_URL, useValue: 'http://localhost:3000'}
  ],
  exports: [
    UserDetailsComponent
  ]
})
export class UserDetailsModule {

  constructor(private translate: TranslateService, private store: Store<language.State>) {
    store.pipe(select(language.getLang)).subscribe((lang: string) => translate.use(lang));
  }

 }
