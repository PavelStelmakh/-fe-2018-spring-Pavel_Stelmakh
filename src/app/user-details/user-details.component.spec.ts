import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { RouterTestingModule } from "@angular/router/testing";
import { HttpClientModule } from '@angular/common/http';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { UserDetailsComponent } from './user-details.component';
import { Observable, of } from 'rxjs';
import { TranslateModule, TranslateLoader, TranslateService } from '@ngx-translate/core';
import { Store } from '@ngrx/store';
import { AuthService } from '../auth/auth.service';
import { UsersService } from '../users.service';
import { BASE_URL } from '../url.service';

const translations = {
  "user-details": {
    "tabOut": "Show Deteils",
    "tabIn": "Edit Deteils",
    "logout": "logout"
  }
};

class FakeLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<{}> {
    return of(translations);
  }
}

describe('UserDetailsComponent', () => {
  let fixture: ComponentFixture<UserDetailsComponent>;
  let translate: TranslateService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        HttpClientModule,
        HttpClientTestingModule,
        TranslateModule.forRoot({
          loader: {provide: TranslateLoader, useClass: FakeLoader}
        })
      ],
      declarations: [
        UserDetailsComponent
      ],
      providers: [
        {provide: BASE_URL, useValue: 'http://localhost:3000'},
        AuthService,
        UsersService,
        {
          provide: Store,
          useValue: {
            dispatch: jest.fn(),
            pipe: jest.fn()
          }
        }
      ],
      schemas: [
       NO_ERRORS_SCHEMA
      ]
    }).compileComponents();

  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserDetailsComponent);
    translate = TestBed.get(TranslateService);
  });

  it('renders markup to snapshot', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

});
