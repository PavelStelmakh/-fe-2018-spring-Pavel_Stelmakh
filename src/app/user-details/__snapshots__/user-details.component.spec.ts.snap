// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetailsComponent renders markup to snapshot 1`] = `
<app-user-details
  auth={[Function AuthService]}
  isEditUsers="false"
  router={[Function Router]}
  select="0"
  store={[Function Object]}
  userService={[Function UsersService]}
>
  <div
    class="tabs"
    ng-reflect-ng-class="[object Object]"
  >
    <div
      class="selectTabFirst select"
      ng-reflect-ng-class="[object Object]"
    >
      user-details.tabOut
    </div>
    
    <div
      class="selectTabThrid"
      ng-reflect-ng-class="[object Object]"
    >
      user-details.tabIn
    </div>
  </div><div
    class="tabsView"
    ng-reflect-ng-switch="0"
  >
    
    <app-output-form />
    
    
  </div><input
    type="button"
    value="user-details.logout"
  />
</app-user-details>
`;
